Bricks/sync/owned_borrowed.h

```diff
@@ -0,0 +1,492 @@
+    destructing_ = true;
+
+    // Collect all the "destructing" callbacks to call.
+    std::vector<std::function<void()>> callbacks_to_call;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      for (const auto& cit : borrowers_) {
+        callbacks_to_call.push_back(cit.second);
+      }
+    }
+
+    // Call all the "destructing" callbacks.
+    for (auto& f : callbacks_to_call) {
+      f();
+    }
+
+    // Wait until all the borrowers have terminated.
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      condition_variable_.wait(lock, [this]() { return borrowers_.empty(); });
+    }
+  }
+
+  // Register ("increase the ref-count") a newly spawned borrower, along with its termination signal callback.
+  // Must be called from within a mutex-locked section.
+  size_t RegisterBorrowerFromLockedSection(std::function<void()> destruction_callback) {
+    borrowers_[++total_borrowers_spawned_throughout_lifetime_] = destruction_callback;
+    return total_borrowers_spawned_throughout_lifetime_;
+  }
+
+  void UpdateBorrowersCallbackFromLockedSection(size_t key, std::function<void()> destruction_callback) {
+    borrowers_[key] = destruction_callback;
+  }
+
+  // Unregister ("decrease the ref-count") a previously spawned borrower.
+  // Locks its own mutex; must be called from outside a mutex-locked section.
+  void UnRegisterBorrower(size_t borrower_index) {
+    bool notify = false;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      auto it = borrowers_.find(borrower_index);
+      CURRENT_ASSERT(it != borrowers_.end());
+      borrowers_.erase(it);
+      notify = borrowers_.empty();
+    }
+    if (notify) {
+      condition_variable_.notify_one();
+    }
+  }
+
+  UniqueInstance(const UniqueInstance&) = delete;
+  UniqueInstance(UniqueInstance&&) = delete;
+  UniqueInstance& operator=(const UniqueInstance&) = delete;
+  UniqueInstance& operator=(UniqueInstance&&) = delete;
+
+  std::atomic_bool destructing_;                        // Whether it already is in the "destructing" mode.
+  T instance_;                                          // The actual instance.
+  std::mutex mutex_;                                    // The mutex to guard per-borrower info.
+  std::condition_variable condition_variable_;          // The variable to notify as all the borrowers are gone.
+  std::map<size_t, std::function<void()>> borrowers_;   // Map of live borrower -> termination signal callback.
+  size_t total_borrowers_spawned_throughout_lifetime_;  // A pre-increment of this value is the key in the above map.
+};
+
+}  // namespace current::sync::impl
+
+template <typename T>
+class Owned;
+template <typename T>
+class Borrowed;
+template <typename T>
+class BorrowedWithCallback;
+template <typename T>
+class BorrowedOfGuaranteedLifetime;
+
+// `WeakBorrowed<T>` is the base class for `Owned<T>`, `Borrowed<T>`, and `BorrowedWithCallback<T>`.
+template <typename T>
+class WeakBorrowed {
```

Comment:
May be it is worth to mention this class in the implementation details above, because it is actually used outside this header, something like:
"The `WeakBorrowed<T>` is the base class for all of them, so it is used if the code doesn't depend on the object ownership details."